### PR TITLE
Get STS to work with MinIO http://localhost:9000

### PR DESCRIPTION
**Pull request by chenkins**

### CODE REVIEW

File: s3/src/main/java/ch/cyberduck/core/s3/S3Session.java [hunk 1]
```diff
@@ -200,14 +197,18 @@ protected RequestEntityRestStorageService connect(final Proxy proxy, final HostK
             configuration.setServiceUnavailableRetryStrategy(new S3TokenExpiredResponseInterceptor(this,
                     new ThreadLocalHostnameDelegatingTrustManager(trust, host.getHostname()), key, prompt));
         }
+        // TODO check with DK: what's the criterion on host to go for AssumeRoleWithWebIdentity?
```
Comment (dkocher) [inline s3/src/main/java/ch/cyberduck/core/s3/S3Session.java:200]: It should be possible to make the decision based on some property in the connection profile.

File: s3/src/main/java/ch/cyberduck/core/s3/S3Session.java [hunk 2]
```diff
@@ -232,8 +233,8 @@ public void login(final Proxy proxy, final LoginCallback prompt, final CancelCal
                         new ThreadLocalHostnameDelegatingTrustManager(trust, host.getHostname()), key, prompt).configure(host);
             }
             // get temporary credentials for MinIO with Web Identity (OIDC)
-            else if (host.getProtocol().getOAuthAuthorizationUrl() != null) {
-                credentials = new NonAwsSTSCredentialsConfigurator(new ThreadLocalHostnameDelegatingTrustManager(trust,
+            else if(host.getProtocol().getOAuthAuthorizationUrl() != null) {
```
Comment (dkocher) [inline s3/src/main/java/ch/cyberduck/core/s3/S3Session.java:236]: The STS endpoint should be configurable in the profile as well and default to `https://sts.amazonaws.com`.

File: s3/src/main/java/ch/cyberduck/core/s3/S3Session.java [hunk 3]
```diff
@@ -243,6 +244,20 @@ else if (host.getProtocol().getOAuthAuthorizationUrl() != null) {
                 client.setProviderCredentials(credentials.isAnonymousLogin() ? null :
                         new AWSSessionCredentials(credentials.getUsername(), credentials.getPassword(),
                                 credentials.getToken()));
+                // TODO remove the following snippet: sanity check that STS credentials are valid with AWS Java SDK
```
Comment (dkocher) [inline s3/src/main/java/ch/cyberduck/core/s3/S3Session.java:247]: Out of scope.

File: s3/src/main/java/ch/cyberduck/core/sts/AssumeRoleWithWebIdentitySTSCredentialsConfigurator.java [hunk 4]
```diff
@@ -16,45 +16,62 @@
...
+                    }
+                })
                 .build();
 
+
         AssumeRoleWithWebIdentityRequest webIdReq = new AssumeRoleWithWebIdentityRequest()
-                .withWebIdentityToken("token")
-                .withPolicy("policy");
+                .withWebIdentityToken(credentials.getOauth().getAccessToken());
+        // TODO check with DK: make configurable in profile/bookmark?
```
Comment (dkocher) [inline s3/src/main/java/ch/cyberduck/core/sts/AssumeRoleWithWebIdentitySTSCredentialsConfigurator.java:70]: Should be configurable in connection profile and fallback to AWS CLI configuration if any.

**Review by dkocher — CHANGES_REQUESTED** [review]

File: profiles/default/S3-OIDC-Testing.cyberduckprofile [hunk 5]
```diff
@@ -20,36 +20,39 @@
...
+        <string>localhost</string>
+        <key>Default Port</key>
+        <string>9000</string>
         <key>Default Nickname</key>
+        <string>MinIO</string>
         <key>Scheme</key>
         <string>http</string>
-        <string>MinIO</string>
         <key>Authorization</key>
-        <string>PasswordGrant</string>
```
Comment (SheromS) [inline profiles/default/S3-OIDC-Testing.cyberduckprofile:33]: This is our testing profile. We tried to use the PasswordGrant flow to test the oAuth flow and the STS service seperately as discussed in the last meeting.

File: s3/src/main/java/ch/cyberduck/core/sts/AssumeRoleWithWebIdentitySTSCredentialsConfigurator.java (hunk 4 above)
Comment (chenkins) [reply s3/src/main/java/ch/cyberduck/core/sts/AssumeRoleWithWebIdentitySTSCredentialsConfigurator.java]: Outdated -> resolve.

File: profiles/default/S3-OIDC-google.cyberduckprofile [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        <true/>
+        <key>Description</key>
+        <string>S3 STS AssumeRoleWithWebIdentity with Google</string>
+        <key>Default Nickname</key>
+        <string>S3 STS</string>
+        <key>Default Hostname</key>
+        <string>s3.amazonaws.com</string>
+        <key>Regions</key>
+        <array>
+            <string>eu-central-1</string>
```
Comment (chenkins) [inline profiles/default/S3-OIDC-google.cyberduckprofile:34]: necessary?

File: s3/src/main/java/ch/cyberduck/core/s3/S3Session.java (hunk 3 above)
Comment (chenkins) [reply s3/src/main/java/ch/cyberduck/core/s3/S3Session.java]: removed snippet -> resolving discussion.

File: s3/src/main/java/ch/cyberduck/core/s3/S3Session.java (hunk 1 above)
Comment (chenkins) [reply s3/src/main/java/ch/cyberduck/core/s3/S3Session.java:200]: @dkocher what would be a good naming convention for the following items and which should we make configurable:
* `stsEndpointURL` -> `.withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(String.format("https://sts.amazonaws.com"), null))`
* `stsTTL`-> `.withDurationSeconds`
*  `stsRole`-> `.withRoleArn("arn:aws:iam::930717317329:role/google-Test-Role")`
* `stsRoleSessionName` -> `.withRoleSessionName("cyberduck-test");`

File: profiles/default/S3-OIDC-google.cyberduckprofile (hunk 6 above)
Comment (dkocher) [reply profiles/default/S3-OIDC-google.cyberduckprofile:34]: Can be omitted to allow all regions from AWS.

File: s3/src/main/java/ch/cyberduck/core/s3/S3Session.java (hunk 1 above)
Comment (dkocher) [reply s3/src/main/java/ch/cyberduck/core/s3/S3Session.java:200]: All lowercase properties with prefix `s3.sts.`.

**Review by IsmCadi — APPROVED** [review]
LGTM - merge with our state and resolving conflicts afterwards

**Pull request closed without merging**